Lib/fontTools/ttLib/woff2.py

```diff
@@ -0,0 +1,1083 @@
+	WOFFFlavorData, sfntDirectoryFormat, sfntDirectorySize, SFNTDirectoryEntry,
+	sfntDirectoryEntrySize, calcChecksum)
+from fontTools.ttLib.tables import ttProgram
+
+haveBrotli = False
+try:
+	import brotli
+	haveBrotli = True
+except ImportError:
+	pass
+
+
+class WOFF2Reader(SFNTReader):
+
+	flavor = "woff2"
+
+	def __init__(self, file, checkChecksums=1, fontNumber=-1):
+		if not haveBrotli:
+			print('The WOFF2 decoder requires the Brotli Python extension, available at:\n'
+				  'https://github.com/google/brotli', file=sys.stderr)
+			raise ImportError("No module named brotli")
+
+		self.file = file
+
+		signature = Tag(self.file.read(4))
+		if signature != b"wOF2":
+			raise TTLibError("Not a WOFF2 font (bad signature)")
+
+		self.file.seek(0)
+		self.DirectoryEntry = WOFF2DirectoryEntry
+		sstruct.unpack(woff2DirectoryFormat, self.file.read(woff2DirectorySize), self)
+
+		self.tables = {}
+		self.tableOrder = []
+		offset = 0
+		for i in range(self.numTables):
+			entry = self.DirectoryEntry()
+			entry.fromFile(self.file)
+			tag = Tag(entry.tag)
+			self.tables[tag] = entry
+			self.tableOrder.append(tag)
+			entry.offset = offset
+			offset += entry.length
+
+		totalUncompressedSize = offset
+		compressedData = self.file.read(self.totalCompressedSize)
+		decompressedData = brotli.decompress(compressedData)
+		if len(decompressedData) != totalUncompressedSize:
+			raise TTLibError(
+				'unexpected size for decompressed font data: expected %d, found %d'
+				% (totalUncompressedSize, len(decompressedData)))
+		self.transformBuffer = StringIO(decompressedData)
+
+		self.file.seek(0, 2)
+		if self.length != self.file.tell():
+			raise TTLibError("reported 'length' doesn't match the actual file size")
+
+		self.flavorData = WOFF2FlavorData(self)
+
+	def __getitem__(self, tag):
+		"""Fetch the raw table data. Reconstruct transformed tables."""
+		tag = Tag(tag)
+		entry = self.tables[tag]
+		rawData = entry.loadData(self.transformBuffer)
+		if tag not in woff2TransformedTableTags:
+			return rawData
+		if hasattr(entry, 'data'):
+			# already reconstructed
+			return entry.data
+		data = self.reconstructTable(tag, rawData)
+		if tag == 'loca' and len(data) != entry.origLength:
+			raise TTLibError(
+				"reconstructed 'loca' table doesn't match original size: expected %d, found %d"
+				% (entry.origLength, len(data)))
+		entry.data = data
+		return entry.data
+
+	def reconstructTable(self, tag, rawData):
+		"""Reconstruct 'glyf' or 'loca' tables from transformed 'rawData'."""
+		if tag not in woff2TransformedTableTags:
+			raise TTLibError("Transform for table '%s' is unknown" % tag)
+		if tag == 'glyf':
+			# reconstruct both glyf and loca
+			self.glyfTable = WOFF2GlyfTable()
+			data = self.glyfTable.reconstruct(rawData)
+		elif tag == 'loca':
+			assert len(rawData) == 0, "expected 0, received %d bytes" % len(rawData)
+			if not hasattr(self, 'glyfTable'):
+				# make sure glyf is loaded first
+				self['glyf']
```

Comment:
Because the transformed loca table is zero bytes, it must be reconstructed alongside the glyf table. The SFNTReader (from which WOFF2Reader inherits) allows to access table data randomly, but a WOFF 2.0 font is meant to be read as a stream, as there are no offsets to specific tables. If the users tries to get "loca" before "glyf", then we must reconstruct the latter as well. But we don't want to do it again and again, so the result is stored as "data" attribute.